src/Server/TCPHandler.cpp

```diff
@@ -873,17 +915,74 @@ void TCPHandler::receiveQuery()
         client_info.initial_query_id = client_info.current_query_id;
         client_info.initial_address = client_info.current_address;
     }
-    else
-    {
-        query_context->setInitialRowPolicy();
-    }
 
     /// Per query settings are also passed via TCP.
     /// We need to check them before applying due to they can violate the settings constraints.
     auto settings_format = (client_revision >= DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS) ? SettingsWriteFormat::STRINGS_WITH_FLAGS
                                                                                                       : SettingsWriteFormat::BINARY;
     Settings passed_settings;
     passed_settings.read(*in, settings_format);
+
+    /// Interserver secret.
+    std::string received_hash;
+    if (client_revision >= DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET)
+    {
+        readStringBinary(received_hash, *in, 32);
+    }
+
+    readVarUInt(stage, *in);
+    state.stage = QueryProcessingStage::Enum(stage);
+
+    readVarUInt(compression, *in);
+    state.compression = static_cast<Protocol::Compression>(compression);
+
+    readStringBinary(state.query, *in);
+
+    /// It is OK to check only when query != INITIAL_QUERY,
+    /// since only in that case the actions will be done.
+    if (!cluster.empty() && client_info.query_kind != ClientInfo::QueryKind::INITIAL_QUERY)
+    {
+#if USE_SSL
+        std::string data(salt);
+        data += cluster_secret;
+        data += state.query;
+        data += state.query_id;
+        data += user_from_hello;
+        data += client_info.initial_user;
+
+        std::string calculated_hash;
+        calculated_hash.resize(32);
+
+        if (received_hash.size() != calculated_hash.size())
+            throw NetException("Unexpected hash received from client", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+
+        encodeSHA256(data, reinterpret_cast<unsigned char *>(calculated_hash.data()));
+
+        if (calculated_hash != received_hash)
+            throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
+        /// TODO: change error code?
+
+        if (query_context->getUserName() != client_info.initial_user)
+        {
+            connection_context.setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
+            query_context->setUserWithoutCheckingPassword(client_info.initial_user, socket().peerAddress());
```

Comment:
Can this be done unconditionally and always, and then only for query_context (need to think about it)

Reply:
It seems so. However in the interserver-secret mode it's necessary to do something about `connection_settings`, which were calculated as the default settings of the user transferred in the hello package. Probably we can just use query settings instead.